### ISSUE TITLE
判别式的机器翻译任务出core

### ISSUE THREAD

**Issue by water1981** [PaddlePaddle/Paddle#907]
现在我们的seq2seq中的机器翻译例子是产生式的（is_generating=True），我想把这个任务改成判别式的（就是只是给出源语言翻译到给定目标语言的得分），训练的时候遵循https://github.com/PaddlePaddle/Paddle/blob/develop/doc/tutorials/text_generation/index_en.md 里面的步骤训练出一个翻译模型， 在test阶段在gen.conf 中设置 is_generating=False ， 结果出core了，错误信息如下：
I /home/img/baidu/idl/paddle/paddle/trainer/Trainer.cpp:148] trainer: in testing mode
I /home/img/baidu/idl/paddle/paddle/trainer/Trainer.cpp:155] trainer mode: Testing
I /home/img/baidu/idl/paddle/paddle/gserver/dataproviders/PyDataProvider2.cpp:247] loading dataprovider dataprovider::process
[INFO 2016-12-15 13:57:33,367 dataprovider.py:27] src dict len : 30000
[INFO 2016-12-15 13:57:33,367 dataprovider.py:37] trg dict len : 30000
I /home/img/baidu/idl/paddle/paddle/gserver/dataproviders/PyDataProvider2.cpp:247] loading dataprovider dataprovider::process
[INFO 2016-12-15 13:57:33,416 dataprovider.py:27] src dict len : 30000
[INFO 2016-12-15 13:57:33,416 dataprovider.py:37] trg dict len : 30000
F /home/img/baidu/idl/paddle/paddle/trainer/Tester.cpp:286] Check failed: (paramUtil_->loadParameters(passId, true , true )) == (true)
/home/img/liushui/bin/paddle: line 81:  4510 Aborted                 (core dumped) ${DEBUGGER} $MYDIR/../opt/paddle/bin/paddle_trainer ${@:2}

**Comment by reyoung** [comment]
因为，得到判别式的模型不是这么设置啊。

而且，似乎整体神经网络的配置也完全不一样了。您能描述一下这个网络究竟应该怎么配置，怎么训练呢？

**Comment by luotao1** [comment]
> 在test阶段在gen.conf 中设置 is_generating=False

test阶段不应该用gen.conf，那是生成的配置。test阶段还是得用训练的配置。

**Comment by water1981** [comment]
训练，我使用的配置是：

is_generating = False

### Data Definiation
data_dir  = "./data/pre-wmt14-bak"
train_conf = seq_to_seq_data(data_dir = data_dir,
                             is_generating = is_generating)

### Algorithm Configuration
settings(
    learning_method = AdamOptimizer(),
    batch_size = 50,
    learning_rate = 5e-4)

### Network Architecture
gru_encoder_decoder(train_conf, is_generating)

**Comment by water1981** [comment]
测试，我使用的配置是：
# whether this config is used for generating
is_generating = False

### Data Definiation
gen_conf = seq_to_seq_data(data_dir = "./data/pre-wmt14",
                           is_generating = is_generating,
                           gen_result = "./translation/gen_result")

### Algorithm Configuration
settings(
      learning_method = AdamOptimizer(),
      batch_size = 1,
      learning_rate = 0)

### Network Architecture
gru_encoder_decoder(gen_conf, is_generating)

**Comment by lcy-seso** [comment]
@water1981 你的需求是不是这样 ： 使用seq2seq model 训练，不生成，而是为两个序列打分，计算 P(Y|X) 这个条件概率呢？X 是源语言序列，Y 是目标语言序列。

如果是上面的需要的话。
1. 使用训练的配置，而不是生成，走 paddle_trainer 的 job=test 流程
2. 配置的Outputs 选择 softmax 。这时会输出每个时间步词典中每一个词是下一个词的概率，需要自己做后处理，把目标词的概率选择出来，logProb 相加，得到整个句子的得分

以上方法有两个缺点：
1. PaddlePaddle 目前木有现成的Layer 能够选择一个向量的某一维输出，上面的方法会令输出文件非常大。
2. softmax 当词表比较大的时候，计算得会比较慢。

其实，如果只是计算两个序列的相似性，下一个词总是知道的，如果可以不去计算全词表中每一个词的概率，那么 score 2 sequence 这个过程是可以极大加速的。

推荐训练的时候使用 PaddlePaddle 的  multi_class_cross_entropy_with_selfnorm 损失函数，预测时使用 SelectiveFc ，可以只输出目标词那一个词的概率，同时避免上面提到的两个问题，极大地加速运算。

**Comment by water1981** [comment]
谢谢，解答！
